src/embind/embind.js

```diff
@@ -405,7 +405,7 @@ var LibraryEmbind = {
   $heap32VectorToArray: function(count, firstElement) {
     var array = [];
     for (var i = 0; i < count; i++) {
-        array.push(HEAP32[(firstElement >> 2) + i]);
+        array.push({{{ makeGetValue('firstElement', 'i * POINTER_SIZE', '*') }}});
```

Comment:
Unless we have special optimizations for this, it will now be emitted as `HEAP32[firstElement + i * POINTER_SIZE >> 2]`, that is, it will use the global `POINTER_SIZE` and do a multiplication, which would be slower than it is now.

We could avoid the global like this:
```suggestion
        array.push({{{ makeGetValue('firstElement', 'i * ' + POINTER_SIZE, '*') }}});
```
The multiplication is more annoying though...

Reply:
Agreed.  Sadly unless we come up with some kind of alliterative to makeGetValue, I don't think we eliminate the multiply.

Reply:
Yes, I think we'd need a new method. Something like `makeArrayLoad(ptr, index, itemSize, offsetInItem)` where we are assumed to be reading an array of items of size `itemSize`, and we are reading the `index`-ed item, and also add a fixed offset inside that item of `offsetInItem`. Then if the memory size is right and the itemSize is right then we could keep the index out of the `>> 2/3` as the current code does.

What do you think about doing that before this PR? Also I could write that function if you prefer.

Reply:
If we do add this new function I think we should keep it as simple as possible (makeGetValue/makeSetValue had become way too complex over the years although they have been cleaned up a long recently).

I'd rather not block this PR on that because:

a) these loops have very few iterations (limited to number of args passed to functions)
b) I think we should be able to find add fix all usages of this pattern (where shift can be hoisted out of loop) and measure the impact the change after this change lands.
c) it could be a slight code size regression to hoist the shift.. but we should measure.

Reply:
d)  we don't have the embind benchmark code checked yet, but once we do we can measure the impact.

Reply:
e) I have more wasm64 embind changes stacked on top of this one.

Reply:
Fair enough. How about at least avoiding usage of the global `POINTER_SIZE`? See the suggested diff in this comment and the other.